src/cargo/util/toml/mod.rs

```diff
@@ -3146,6 +3157,120 @@ impl<'de> de::Deserialize<'de> for TomlDebugInfo {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd, Hash, Serialize)]
+#[serde(untagged)]
+pub enum TomlTrimPaths {
+    One(TomlTrimPathsValue),
+    Many(Vec<TomlTrimPathsValue>),
+}
+
+impl<'de> de::Deserialize<'de> for TomlTrimPaths {
+    fn deserialize<D>(d: D) -> Result<TomlTrimPaths, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let expecting = r#"a boolean, "none", "macro", "diagnostics", "object", "all", or a list of these options"#;
+        UntaggedEnumVisitor::new()
+            .expecting(expecting)
+            .bool(|value| {
+                Ok(if value {
+                    TomlTrimPaths::all()
+                } else {
+                    TomlTrimPaths::none()
+                })
+            })
+            .string(|value| {
+                TomlTrimPathsValue::deserialize(value.into_deserializer()).map(TomlTrimPaths::One)
+            })
+            .seq(|seq| {
+                let seq: Vec<String> = seq.deserialize()?;
+                let seq = seq
+                    .into_iter()
+                    .map(|s| TomlTrimPathsValue::deserialize(s.into_deserializer()))
+                    .collect::<Result<_, _>>()?;
+                Ok(TomlTrimPaths::Many(seq))
+            })
+            .deserialize(d)
+    }
+}
+
+impl TomlTrimPaths {
+    pub fn none() -> TomlTrimPaths {
+        TomlTrimPaths::One(TomlTrimPathsValue::None)
+    }
+
+    pub fn all() -> TomlTrimPaths {
+        TomlTrimPaths::One(TomlTrimPathsValue::All)
+    }
+
+    pub fn diagnostics() -> TomlTrimPaths {
+        TomlTrimPaths::One(TomlTrimPathsValue::Diagnostics)
+    }
+
+    pub fn r#macro() -> TomlTrimPaths {
+        TomlTrimPaths::One(TomlTrimPathsValue::Macro)
+    }
+
+    pub fn object() -> TomlTrimPaths {
+        TomlTrimPaths::One(TomlTrimPathsValue::Object)
+    }
+
+    pub fn is_none(&self) -> bool {
+        match self {
+            TomlTrimPaths::One(TomlTrimPathsValue::None) => true,
+            TomlTrimPaths::Many(list) if list.is_empty() => true,
+            TomlTrimPaths::Many(list) => list.iter().any(|v| matches!(v, TomlTrimPathsValue::None)),
+            _ => false,
+        }
+    }
+}
+
+impl fmt::Display for TomlTrimPaths {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            TomlTrimPaths::One(value) => write!(f, "{value}"),
+            TomlTrimPaths::Many(values) => {
+                let mut iter = values.iter();
+                if let Some(value) = iter.next() {
+                    write!(f, "{value}")?;
+                }
+                for value in iter {
+                    write!(f, ",{value}")?;
+                }
+                Ok(())
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum TomlTrimPathsValue {
+    None,
```

Comment:
Do we need a `None` or should we use the empty list?

Reply:
Started writing this up but realized I was thinking of the wrong stuff, so sharing it here if its relevant

`package.publish` does this kind of things though in post-processing (`to_real_manifest`).  Its a `VecOrBool` and the `false` gets turned into `Some(vec![])`, `true` gets mapped to `None` (ie all), and then `Some(vec![...])` is the enumerated items.

Reply:
Just pushed a change disallowing empty array during deserialization.

Reply:
How come, doesn't an empty list communicate the same thing?  The empty list would also parallel `package.publish`

Reply:
It is not really straightforward. An empty array is not always falsy in programming world, and it seems better to reduce the size of possible options.

Anyway, given we have the existing pattern, I'll follow.